[IRGen] Adjust element size of offset vector to 32-bit for structs

xedin opened the pull request:
Type of elements contained by field offsets vector can be adjusted
to 32-bit integers (from being pointer sized) to safe space in the
binary since segment size is limited to 4 GB.

Resolves: rdar://problem/36560486

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin commented on lib/IRGen/GenMeta.cpp:
```diff
@@ -2745,10 +2745,11 @@ namespace {
         }
       };
       Scanner(IGM, Target, getLoweredType(), init).layout();
+      init.addAlignmentPadding(IGM.getPointerAlignment());
```
@jckarter I'm facing a problem with this `addAlignmentPadding` since it adds `[4 x i8] zero initialized array` as an element which differs from of array element type (which results in failure in `finishArray`), is the a workaround for this? I guess there should be another way to get the end address of the array after it has been finalized, but i can't find one...

jckarter replied:
`addAlignmentPadding` probably ought to be overridden by `ConstantArrayBuilder` to form alignment padding using the array element type instead of byte arrays.

xedin commented:
@swift-ci please smoke test

xedin commented:
@jckarter I think I'm getting really close there is only one real failure now in "key path generic instantiation", I think it's related to changes in `GenericStructMetadataBuilder`, I'll dig into it some more tomorrow.

xedin committed:
[IRGen] Adjust element size of offset vector to 32-bit for structs

Type of elements contained by field offsets vector can be adjusted
to 32-bit integers (from being pointer sized) to safe space in the
binary since segment size is limited to 4 GB.

Resolves: rdar://problem/36560486

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin committed:
[IRGen] NFC: Fix tests to reflect struct field vector size changes

xedin commented:
@swift-ci smoke test compiler performance

xedin commented:
@swift-ci please smoke test compiler performance

jckarter reviewed (approved):
Looks good, thanks!

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test source compatibility
